### ISSUE TITLE
udp.Port schedules connectionLost call in the global reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5299]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)| @zseil reported|
|-|-|
|Trac ID|trac#5299|
|Type|defect|
|Created|2011-10-09 19:26:14Z|

`udp.Port._loseConnection` imports the global reactor to `callLater` its `connectionLost` method. It should use `self.reactor` instead.

Attachments:

* [udp-stopProtocol-scheduling-5299.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0a9/0a99f44bcfa3f54e543b939e44ad00144beff930/3b41fc3397af2a7bb7f4c06fec468266b3bb7005.patch) (5037 bytes) - added by zseil on 2011-10-09 20:41:00Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5299 5299
type__defect defect
reporter__zseil zseil
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1318188374000000 1318188374000000
changetime__1318794115000000 1318794115000000
version__None None
owner__zseil zseil

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_1"></a>|@zseil commented|
|-|-|

Here is a patch that should fix this issue. The change to `iocpreactor.upd.Port.connectionLost` could be smaller, it is only missing a `self.reactor.removeActiveHandle` call, but this way it is more clearly visible that the whole method is copy-pasted. I also removed `connectUDP` leftovers that were missed in [#4075](https://github.com/twisted/twisted/issues/4075).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @zseil|
|-|-|

That looks pretty good, with a couple minor changes:

   1. The way you test whether the stopProtocol is synchronous is a bit confusing. Perhaps less confusing would be to have `startProtocol` do `self.inStartProtocol = True`, stop the port, `self.inStartProtocol = False`. Then you can just assert in stopProtocol that `self.inStartProtocol` is False.
   1. The unrelated changes should in a different ticket, if they're unrelated.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [32870]) Add tests for the UDP port start and stop log messages; make the stop message include "UDP"; also fix the UDP stopListening implementation to not use the global reactor

refs [#5308](https://github.com/twisted/twisted/issues/5308)
refs [#5299](https://github.com/twisted/twisted/issues/5299)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [32871]) Replicate those fixes exactly in iocpreactor

refs [#5308](https://github.com/twisted/twisted/issues/5308)
refs [#5299](https://github.com/twisted/twisted/issues/5299)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [32872]) Incorporate zseil's test from [[#5299](https://github.com/twisted/twisted/issues/5299)](https://github.com/twisted/twisted/issues/5299)

refs [#5308](https://github.com/twisted/twisted/issues/5308)
refs #5299

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

I incorporated this change into the [#5308](https://github.com/twisted/twisted/issues/5308) branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [32883]) Merge udp-port-ident-5308

Author: zseil, exarkun
Reviewer: itamar, jesstess
Fixes: [#5299](https://github.com/twisted/twisted/issues/5299)
Fixes: [#5308](https://github.com/twisted/twisted/issues/5308)

Change UDP ports to:

  * Use the `DatagramProtocol`'s `ILoggingContext` implementation for startup messages
  * log their type (ie, UDP) when they are closed
  * Use the reactor passed to their initializer to schedule connectionLost, rather than the global reactor
